### PR TITLE
Implement ranges::elements_view

### PR DESCRIPTION
**Pull request by miscco**
This is not fully functional as I has hit a roadblock, trying to understand why `ranges::view<EV>` breaks due to `ranges::end` not being available.

@CaseyCarter: any ideas what is going on here?

### CODE REVIEW

File: stl/inc/ranges [hunk 1]
```diff
@@ -2059,6 +2059,364 @@ namespace ranges {
...
+
+    // CLASS TEMPLATE ranges::elements_view
+    // CONCEPT _Has_tuple_element
+    template <class _Tuple, size_t _Index>
+    concept _Has_tuple_element = requires(_Tuple __t) {
+        typename tuple_size<_Tuple>::type;
+        requires _Index < tuple_size_v<_Tuple>;
+        typename tuple_element_t<_Index, _Tuple>;
+        // clang-format off
+        { get<_Index>(__t) } -> convertible_to<const tuple_element_t<_Index, _Tuple>&>;
```
Comment (cpplearner) [inline stl/inc/ranges:2071]: Should this block ADL?

```suggestion
        { _STD get<_Index>(__t) } -> convertible_to<const tuple_element_t<_Index, _Tuple>&>;

```

Comment (miscco) [reply stl/inc/ranges]: Absolutly, that one has slipped by :+1:

File: stl/inc/ranges [hunk 2]
```diff
@@ -2059,6 +2059,365 @@ namespace ranges {
...
+
+        template <class _Traits, class _Base> // TRANSITION, LWG-3289
+        struct _Iterator_base {};
+
+        // clang-format off
+        template <class _Traits, class _Base>
+            requires _Has_member_iterator_category<_Traits>
+        struct _Iterator_base<_Traits, _Base> {
+            // clang-format on
+            using iterator_category = typename _Traits::iterator_category;
```
Comment (miscco) [inline stl/inc/ranges:2096]: Is that actually necessary, I have stolen it from transform_view which does a quite a bit of work here.

**Comment by cpplearner** [comment]
Note that `elements_view` as specified in the current WD is broken. See LWG-3406.

**Comment by miscco** [comment]
> Note that `elements_view` as specified in the current WD is broken. See [LWG-3406](https://cplusplus.github.io/LWG/issue3406).

Awesome, thanks for the heads up 😺

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/ranges [hunk 3]
```diff
@@ -2059,6 +2059,379 @@ namespace ranges {
...
+            friend class _Iterator;
+            template <bool>
+            friend class _Sentinel;
+
+            using _Base = _Maybe_const<_Const, _Vw>;
+
+            iterator_t<_Base> _Current{};
+
+        public:
+            using iterator_category = typename iterator_traits<iterator_t<_Base>>::iterator_category;
```
Comment (cpplearner) [inline stl/inc/ranges:2112]: `iterator_category` is inherited from `_Iterator_base`.

```suggestion
```

**Review by CaseyCarter — CHANGES_REQUESTED** [review]
Changes look good - I think we're only missing some conditional `noexcept` test coverage now (https://github.com/microsoft/STL/pull/1406#discussion_r513011605).

File: tests/std/tests/P0896R4_views_elements/test.cpp [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+    STATIC_ASSERT(CanMemberSize<R> == CanSize<V>);
+    STATIC_ASSERT(CanSize<R> == CanMemberSize<R>);
+    STATIC_ASSERT(CanMemberSize<const R> == CanSize<const V>);
+    STATIC_ASSERT(CanSize<const R> == CanMemberSize<const R>);
+    if constexpr (CanMemberSize<R>) {
+        assert(r.size() == static_cast<decltype(r.size())>(ranges::size(expected_keys)));
+        if constexpr (CanMemberSize<const R>) {
+            assert(as_const(r).size() == static_cast<decltype(r.size())>(ranges::size(expected_keys)));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_views_elements/test.cpp:166]: #1409.

**Review by CaseyCarter — APPROVED** [review]
> Changes look good - I think we're only missing some conditional `noexcept` test coverage now ([#1406 (comment)](https://github.com/microsoft/STL/pull/1406#discussion_r513011605)).

Nice try, but you were wrong. Fixed up a few small issues and I think this is now good to go.

File: stl/inc/ranges [hunk 5]
```diff
@@ -2059,6 +2059,382 @@ namespace ranges {
...
+            using iterator_concept = conditional_t<random_access_range<_Vw>, random_access_iterator_tag,
+                conditional_t<bidirectional_range<_Vw>, bidirectional_iterator_tag,
+                    conditional_t<forward_range<_Vw>, forward_iterator_tag, input_iterator_tag>>>;
+            using value_type       = remove_cvref_t<tuple_element_t<_Index, range_value_t<_Base>>>;
+            using difference_type  = range_difference_t<_Base>;
+
+            _Iterator() = default;
+
+            constexpr explicit _Iterator(iterator_t<_Base> _Current_) noexcept(
+                noexcept(is_nothrow_move_constructible_v<iterator_t<_Base>>)) // strengthened
```
Comment (StephanTLavavej) [inline stl/inc/ranges:2119]: Bug: `noexcept(noexcept(is_nothrow_move_constructible_v<iterator_t<_Base>>))` is always true.

It appears that this was copy-pasted from `transform_view`; I created #1417 to fix the pre-existing occurrences.

**Review by StephanTLavavej — COMMENTED** [review]
(This is a very incomplete review; I'll resume reviewing later this week)

File: stl/inc/ranges [hunk 6]
```diff
@@ -2059,6 +2059,382 @@ namespace ranges {
...
+            friend class _Sentinel;
+
+            using _Base = _Maybe_const<_Const, _Vw>;
+
+            iterator_t<_Base> _Current{};
+
+        public:
+            using iterator_concept = conditional_t<random_access_range<_Vw>, random_access_iterator_tag,
+                conditional_t<bidirectional_range<_Vw>, bidirectional_iterator_tag,
+                    conditional_t<forward_range<_Vw>, forward_iterator_tag, input_iterator_tag>>>;
```
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/ranges:2112]: This is necessary for the `random_access_range` concepts to work. As far as I can tell this is what LWG-3289 is about

[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/ranges:2112]: I see. In that case, I have no objection to keeping this.

File: stl/inc/ranges (hunk 5 above)
Comment (miscco) [reply stl/inc/ranges:2119]: Thanks, fixed

File: stl/inc/ranges (hunk 6 above)
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/ranges:2112]: I believe the whole ``_Iterator_base` hack is simply for LWG-3289, so I guess we should keep it as it is?

Asking @CaseyCarter for verification

[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/ranges:2112]: Yes, this is the approach we've been using to deal with LWG-3289 for the time being, so our adaptors can adapt ranges with move-only iterators which have no corresponding C++17 iterator category.

File: stl/inc/ranges [hunk 7]
```diff
@@ -2059,6 +2059,382 @@ namespace ranges {
...
+    class elements_view : public view_interface<elements_view<_Vw, _Index>> {
+        // clang-format on
+    private:
+        /* [[no_unique_address]] */ _Vw _Range{};
+
+        template <bool _Const>
+        class _Sentinel;
+
+        template <class _Traits, class _Base> // TRANSITION, LWG-3289
+        struct _Iterator_base {};
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/ranges:2087]: I'm preparing a fix for #1418 renaming these types to `_Category_base`, which I think is an improvement regardless of shadowing. I'll push changes here as well.

File: tests/std/test.lst [hunk 8]
```diff
@@ -341,6 +341,7 @@ tests\P0896R4_stream_iterators
 tests\P0896R4_views_all
 tests\P0896R4_views_drop
 tests\P0896R4_views_empty
+tests\P0896R4_views_elements
```
Comment (StephanTLavavej) [inline tests/std/test.lst:344]: Sort:
```suggestion
tests\P0896R4_views_elements
tests\P0896R4_views_empty
```

File: tests/std/tests/P0896R4_views_elements/test.cpp [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+
+    // Validate elements_view::base() const&
+    STATIC_ASSERT(CanMemberBase<const R&> == copy_constructible<V>);
+    if constexpr (copy_constructible<V> && forward_range<V>) {
+        same_as<V> auto b1 = as_const(r).base();
+        STATIC_ASSERT(noexcept(as_const(r).base()) == is_nothrow_copy_constructible_v<V>);
+        if (!is_empty) {
+            assert(*b1.begin() == pair(0, -1)); // NB: depends on the test data
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_elements/test.cpp:299]: Style: Would prefer to see braces for temporary pairs, to make them look less like function calls. Occurs 4 times total.

**Review by StephanTLavavej — COMMENTED** [review]
This looks great, thanks! All I found were extremely minor nitpicks so I'll validate and push changes.

**Review by StephanTLavavej — APPROVED** [review]
@CaseyCarter I pushed small changes after you approved.

**Review by CaseyCarter — APPROVED** [review]

**Comment by CaseyCarter** [comment]
> @CaseyCarter I pushed small changes after you approved.

@StephanTLavavej I pushed what look like bug changes after you approved: `common_view` and `elements_view` conflicted by being at the same point in `<ranges>`, which I resolved by relocating `common_view` before `reverse_view` in proper synopsis order. There are no other changes in that commit, it's just a cut-n-paste job.

**Comment by CaseyCarter** [comment]
> @StephanTLavavej I pushed what look like bug changes after you approved: `common_view` and `elements_view` conflicted by

Err, "big changes". This is too funny to edit in place, I'm leaving the typo for posterity ;)

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
FYI @CaseyCarter, I've pushed a fix after you approved, working around an internal compiler assertion in the test that I've reduced and reported.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
* git automatically merged `<ranges>`; GH-1366 added `drop_while_view` next to where this PR moved `common_view`. I double-checked that they're in that order (synopsis order).
* Fixed trivial merge conflicts:
  + In `tests/std/test.lst`, GH-1366 added `P0896R4_views_drop_while[_death]` and this PR added `P0896R4_views_elements`. Accepted both, in that order (sorted order).
  + In `P0896R4_ranges_range_machinery/test.cpp`, GH-1366 added a line testing `drop_while` and this PR added a line testing `elements<42>`. Accepted both, in that order (sorted order).

**Comment by StephanTLavavej** [comment]
Thanks for implementing these crucial **elements** of ranges; customers will be able to **view** them in VS 2019 16.9 Preview 2. :joy_cat:

**Pull request merged**